Add Support for UserFields in ZendeskApi_v2

ruannbotes opened the pull request:
Summary:
This pull request adds support for managing UserFields within the ZendeskApi_v2 package. It includes both synchronous and asynchronous methods for CRUD operations on UserFields. The implementation aligns with the existing structure for managing TicketFields, ensuring consistency and ease of use.

Changes Include:

API Methods: Implemented sync and async methods for UserFields operations.
Models: Added UserField, GroupUserFieldResponse, and IndividualUserFieldResponse models to represent UserFields and responses.
Constants: Introduced UserFieldTypes constants for standardized handling of field types.
Requests Update: Extended Users requests to support the new UserFields functionality.
Unit Tests: Updated UsersTests to cover the new UserField methods, including both sync and async operations.
Testing:

Added comprehensive unit tests to ensure proper functionality for all new methods.
Tests include validation of input and output for the UserFields operations, ensuring they work as expected.

ruannbotes committed:
feat: add UserFieldTypes constants

- Introduced constants for UserFieldTypes to ensure consistent usage across the codebase.

ruannbotes committed:
feat: add models for UserField responses and data structures

- Created models: GroupUserFieldResponse, IndividualUserFieldResponse, UserField.
- These models handle the structure of data returned by UserField endpoints.

ruannbotes committed:
feat: update Users requests to support UserField operations

- Implemented sync and async methods to handle UserField endpoints.

ruannbotes committed:
test: implement unit tests for UserField functionality

- Added tests for sync and async UserField API methods.
- Updated UsersTests to cover new UserField scenarios.

ruannbotes committed:
fix: correct parameter names from ticketField to userField in async interface methods

Corrected the parameter names in the async interface methods where ticketField was mistakenly used instead of userField.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/Speedygeek/ZendeskApi_v2?pullRequest=599) <br/>All committers have signed the CLA.

ruannbotes commented:
Hi @mozts2005,

I can address the issues flagged by CodeFactor. However, the test failures in the pipeline seem to be related to invalid or expired secrets. The pipeline output shows the following error:

```
jsonException: {"error":"Couldn't authenticate you"}
2024-08-20T11:47:02.1169102Z ----> System.Net.WebException : The remote server returned an error: (401) Unauthorized.
```

It looks like the authentication details may need to be updated. Could you confirm if the secrets used in the pipeline are still valid?

Thanks!

mozts2005 commented:
Yes, I will update the pipeline secret and rerun the test.

ruannbotes committed:
Fix: Refactor UserField Classes and Update Tests

- Separated UserField classes into individual files.
- Consolidated fields into a unified UserField class.
- Moved CustomFieldOption to a shared folder for reuse.
- Updated UserTests.cs to reflect structural changes.

ruannbotes committed:
Fix: remove the space

-An opening brace should not be followed by a blank line. (SA1505)

ruannbotes commented:
Hi @mozts2005,

I have resolved the CodeFactor issues, tests still failing due to authentication which I believe will be resolved with your secret update.

Thanks!

ruannbotes commented:
Hi @mozts2005 

Hope you doing well, just following on the pull request and pipeline secrets issue.

Thanks.

MelGrubb commented:
Is there any movement on this? Or maybe a fork somewhere that has this already? I am stuck needing to update a user field and there's no support in the current version of the API. I either have to fork my own and add it myself following this PR, or I'm going to have to go raw http-request.

I'd really rather not
a) Take on responsibility for yet another library
OR
b) Introduce another inconsistency by using a library for "all but one" call that I need to make to Zendesk.
